torch/csrc/autograd/custom_function.h

```diff
@@ -173,6 +180,55 @@ struct CppNode : public Node {
 
   void set_ctx_grad_fn(const std::shared_ptr<Node>& node);
   void save_variables_to_ctx();
+
+  void compiled_args(CompiledNodeArgs& args) override {
+    if (!T::is_traceable) {
+      throw std::runtime_error(
+          std::string(
+              "compiled_args not implemented for non-traceable node: ") +
+          name());
+    }
+
+    // although neither of the 2 methods below have uniqueness guarantees
+    // it is unlikely for them to collide at the same time
+    args.collect(typeid(T).hash_code());
+    args.collect(std::string(typeid(T).name()));
```

Comment:
I believe we already specialize on the type of the node here:
https://github.com/pytorch/pytorch/blob/8e4301077efea3265d6707735df5fa56757f6105/torch/csrc/dynamo/compiled_autograd.h#L42-L44

Reply:
I used that initially and it failed the "test_autograd_cpp_node_id" test due to `typeid(CppNode<T1>) == typeid(CppNode<T2>)`. It happens consistently, so I think we need to additionally specialize on the template type

Reply:
I see.